vignettes/converting_files.Rmd

```diff
@@ -0,0 +1,97 @@
+---
+title: "Converting from XPT"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Converting from XPT}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+library(datasetjson)
+```
+
+Some users may be interested in converting SAS Version 5 Transport files into XPT, or converting from other file types such as SAS7BDAT. There may be some existing processes in place to do this in SAS, such as a post-processing conversion step where files are converted in bulk. This vignette offers some guidance and best practice for doing this process in R. 
+
+## Converting from XPT
+
+Ideally if you're converting from an XPT file or a SAS7BDAT file, the appropriate metadata would be maintained separate from the dataset itself. Inevitably some information will get lost while being exchanged because of differences of what information was applied or how R interprets SAS datasets. This is one reason why **{datasetjson}** improves upon XPTs as an exchange format. In leiu of external metadata, here's an example of how you can make a best effort conversion.
+
+```{r XPT conversion}
+adsl <- haven::read_xpt(file.path(system.file(package='datasetjson'), "adsl.xpt"))
+
+#' Gather variable metadata in Dataset JSON compliant format
+#'
+#' @param n Variable name
+#' @param .data Dataset to gather attributes
+#'
+#' @returns Columns compliant data frame
+extract_xpt_meta <- function(n, .data) {
+  
+  attrs <- attributes(.data[[n]])
+  
+  out <- list()
+
+  # Identify the variable type
+  if (inherits(.data[[n]],"Date")) {
+    out$dataType <- "date"
+    out$targetDataType <- "integer"
+  } else if (inherits(.data[[n]],"POSIXt")) {
+    out$dataType <- "datetime"
+    out$targetDataType <- "integer"
+  } else if (inherits(.data[[n]],"numeric")) {
+    if (any(is.double(.data[[n]]))) out$dataType <- "float"
+    else out$dataType <- "integer"
+  } else {
+    out$dataType <- "string"
+    out$length <- max(purrr::map_int(.data[[n]], nchar))
+  }
+  
+  out$itemOID <- n
+  out$name <- n
+  out$label <- attr(.data[[n]], 'label')
+  out$displayFormat <- attr(.data[[n]], 'format.sas')
+  tibble::as_tibble(out)
+  
+}
+
+# Loop the ADSL columns
+adsl_meta <- purrr::map_df(names(adsl), extract_xpt_meta, .data=adsl)
```

Comment:
I think we should chuck this out so we can print the objects and show/explain more what is happening here.  If you agree, I'm happy to make some tweaks.

Reply:
So keep it as a list? If so, the extract_xpt_meta function returns a tibble. 

I had it as a data frame because that's typically what we use to provide the metadata into the `dataset_json()` function. But happy for you to expand it.